File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-3.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    goodNodes(root) {
        let res = []
        this.dfs(root, Number.MIN_SAFE_INTEGER, res)
        return res.length
    }

    dfs(root, max, res){
        if (!root) {return}

        if (max <= root.val) {res.push(root.val)}
        max = Math.max(max, root.val)

        this.dfs(root.left, max, res)
        this.dfs(root.right, max, res)
    }
}

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-3.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func lowestCommonAncestor(root *TreeNode, p *TreeNode, q *TreeNode) *TreeNode {
    if root.Val < q.Val && root.Val < p.Val {
		return lowestCommonAncestor(root.Right, p, q)
	} else if root.Val > q.Val && root.Val > p.Val {
		return lowestCommonAncestor(root.Left, p, q)
	} else {
		return root
	}
}
